python/cudf/cudf/dataframe/multiindex.py

```diff
@@ -204,76 +204,102 @@ def _compute_levels_and_codes(self):
+                if row not in index.levels[idx]:
+                    raise KeyError(row)
+        return result
 
-    def _get_row_major(self, df, row_tuple):
-        slice_access = False
-        if isinstance(row_tuple[0], numbers.Number):
-            valid_indices = row_tuple[0]
-        elif isinstance(row_tuple[0], slice):
-            # 1. empty slice compute
-            if row_tuple[0].stop == 0:
-                valid_indices = []
-            else:
-                slice_access = True
-                start = row_tuple[0].start or 0
-                stop = row_tuple[0].stop or len(df)
-                step = row_tuple[0].step or 1
-                valid_indices = cudautils.arange(start, stop, step)
-        else:
-            valid_indices = self._compute_validity_mask(df, row_tuple)
+    def _get_valid_indices_by_tuple(self, index, row_tuple, max_length):
+        from cudf.utils.cudautils import arange
         from cudf import Series
 
-        result = df.take(Series(valid_indices))
-        # Build new index - INDEX based MultiIndex
-        # ---------------
+        # Instructions for Slicing
+        # if tuple, get first and last elements of tuple
+        # if open beginning tuple, get 0 to highest valid_index
+        # if open ending tuple, get highest valid_index to len()
+        # if not open end or beginning, get range lowest beginning index
+        # to highest ending index
+        if isinstance(row_tuple, slice):
+            if (
+                isinstance(row_tuple.start, numbers.Number)
+                or isinstance(row_tuple.stop, numbers.Number)
+                or row_tuple == slice(None)
+            ):
+                stop = row_tuple.stop or max_length
+                start, stop, step = row_tuple.indices(stop)
+                return arange(start, stop, step)
+            start_values = self._compute_validity_mask(
+                index, row_tuple.start, max_length
+            )
+            stop_values = self._compute_validity_mask(
+                index, row_tuple.stop, max_length
+            )
+            return Series(arange(start_values.min(), stop_values.max() + 1))
+        elif isinstance(row_tuple, numbers.Number):
+            return row_tuple
+        return self._compute_validity_mask(index, row_tuple, max_length)
+
+    def _index_and_downcast(self, result, index, index_key):
         from cudf import DataFrame
+        from cudf import Series
+
+        if isinstance(index_key, (numbers.Number, slice)):
+            index_key = [index_key]
+        if (
+            len(index_key) > 0 and not isinstance(index_key, tuple)
+        ) or isinstance(index_key[0], slice):
+            index_key = index_key[0]
 
+        slice_access = False
+        if isinstance(index_key, slice):
+            slice_access = True
         out_index = DataFrame()
         # Select the last n-k columns where n is the number of source
         # levels and k is the length of the indexing tuple
```

Comment:
Stale comment?

Reply:
They're not exactly stale, but obviously don't make much sense! I'm looking to see if I can rewrite them in a way that adds value. That was the intention.